presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -539,15 +626,61 @@ public Duration getTotalCpuTime()
     public void start()
     {
         if (started.compareAndSet(false, true)) {
-            executor.submit(this::schedule);
+            startScheduling(getStagesReadyForExecution());
+        }
+    }
+
+    private List<SqlStageExecution> getStagesReadyForExecution()
+    {
+        return stream(forTree(StreamingPlanSection::getChildren).depthFirstPreOrder(sectionedPlan))
+                .filter(this::isReadyForExecution)
+                .flatMap(section -> stream(forTree(SubPlan::getChildren).depthFirstPreOrder(section.getPlan())))
+                .map(SubPlan::getFragment)
+                .map(PlanFragment::getId)
+                .map(this::getFragmentExecution)
+                .collect(toImmutableList());
+    }
+
+    private boolean isReadyForExecution(StreamingPlanSection section)
+    {
+        SqlStageExecution stage = getFragmentExecution(section.getPlan().getFragment().getId());
+        if (stage.getState() != PLANNED) {
+            return false;
         }
+        for (StreamingPlanSection child : section.getChildren()) {
+            SqlStageExecution childRootStage = getFragmentExecution(child.getPlan().getFragment().getId());
+            if (childRootStage.getState() != FINISHED) {
+                return false;
+            }
+        }
+        return true;
     }
 
-    private void schedule()
+    private SqlStageExecution getFragmentExecution(PlanFragmentId planFragmentId)
     {
+        return stages.get(planFragmentIdToStageIdMap.get(planFragmentId));
+    }
+
+    private void startScheduling(Collection<SqlStageExecution> stages)
```

Comment:
nit: what about call it `doScheduling`?

Reply:
i like startScheduling

Reply:
I would also prefer to keep it `startScheduling`. As this method doesn't do the scheduling, but just starts it in the background.